log/DESIGN.md

```diff
@@ -0,0 +1,191 @@
+
+### Record
+
+The [`LogRecord` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#logger)
+is defined as a struct.
+
+```go
+type Record struct {
+	Timestamp         time.Time
+	ObservedTimestamp time.Time
+	Severity          Severity
+	SeverityText      string
+	Body              string
+
+	// Allocation optimization: an inline array sized to hold
+	// the majority of log calls (based on examination of open-source
+	// code). It holds the start of the list of Attrs.
+	front [nAttrsInline]attribute.KeyValue
+
+	// The number of Attrs in front.
+	nFront int
+
+	// The list of Attrs except for those in front.
+	// Invariants:
+	//   - len(back) > 0 iff nFront == len(front)
+	//   - Unused array elements are zero. Used to detect mistakes.
+	back []attribute.KeyValue
+}
+
+const nAttrsInline = 5
+
+type Severity int
+
+const (
+	SeverityUndefined Severity = iota
+	SeverityTrace
+	SeverityTrace2
+	SeverityTrace3
+	SeverityTrace4
+	SeverityDebug
+	SeverityDebug2
+	SeverityDebug3
+	SeverityDebug4
+	SeverityInfo
+	SeverityInfo2
+	SeverityInfo3
+	SeverityInfo4
+	SeverityWarn
+	SeverityWarn2
+	SeverityWarn3
+	SeverityWarn4
+	SeverityError
+	SeverityError2
+	SeverityError3
+	SeverityError4
+	SeverityFatal
+	SeverityFatal2
+	SeverityFatal3
+	SeverityFatal4
+)
+```
+
+`Record` has `Attributes` and `AddAttributes` methods,
+like [`slog.Record.Attrs`](https://pkg.go.dev/log/slog#Record.Attrs)
+and [`slog.Record.AddAttrs`](https://pkg.go.dev/log/slog#Record.AddAttrs),
+in order to achieve high-performance when accessing and setting attributes efficiently.
+
+The `NewRecord(...RecordOption) (Record, error)` is a factory function
+used to create records using provided options.
+
+`Record` has a `Clone` method to allow copying records
+so that the SDK can offer concurrency safety.
+
+## Compatibility
+
+The backwards compatibility is achieved using the `embedded` design pattern
+that is already used in Trace API and Metrics API.
+
+## Benchmarking
+
+The benchmarks takes inspiration from [`slog`](https://pkg.go.dev/log/slog),
+because for the Go team it was also critical to create API that would be fast
+and interoperable with existing logging packages.[^1]
+
+## Rationale
+
+### Logger.Emit definition
+
+One of the ideas was to have:
+
+```go
+type Logger interface{
+    Emit(ctx context.Context, record Record)
+}
+```
+
+This gives the advantage that the SDK would not need to call `NewRecord(options...)`.
+
+The user can still easily create a helper that could be easier to use:
+
+```go
+func log(ctx context.Context, l Logger, options ...RecordOption) {
+    r := log.NewRecord(options...)
+    l.Emit(ctx, r)
+}
+```
+
+The main reasons against this definition are that following:
+
+1. The existing design is similar to the [Meter API](https://pkg.go.dev/go.opentelemetry.io/otel/metric#Meter)
+for creating instruments.
+2. It is unsure if anyone would like to reuse a record.
+3. Just passing options should be more-user friendly API.
```

Comment:
I like the option pattern (we would likely still need options if we went this route), but having the Record in an option has some downsides for UX:

* What happens if a user passes two options with log records?  Do they both get emitted?  Or just one?
* What happens if a user doesn't pass any log record option?
* Extra WithLogRecord() words:

```golang
logger.Emit(ctx, log.Record{Body: "bar"})
```
vs 
```golang
logger.Emit(ctx, log.WithLogRecord(log.Record{Body: "bar"}))
```

It reminds me a bit of Observe() from the metrics api: https://github.com/open-telemetry/opentelemetry-go/blob/47ba653e6962948dab3a481231c469387f11f184/metric/asyncfloat64.go#L223

We could have made `value` a `WithValue` option, but it is almost always wrong for users to pass more than one value, or no value, so it was better as an explicit argument to the function.

Reply:
I was planning to add option: `WithBody`, `WithTimestamp`. I do NOT want to add a `WithRecord` option. The main reason is validation. I want the SDK to take care of the validation and I think the easiest way would be to simply return error from `NewRecord`.

Reply:
Makes sense.  Would NewRecord be private then?

Reply:
Ah, right. Looks good.